packages/eslint-plugin/src/rules/no-array-delete.ts

```diff
@@ -0,0 +1,57 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import * as ts from 'typescript';
+
+import * as util from '../util';
+
+export default util.createRule<[], 'arrayDeleteViolation'>({
+  name: 'no-array-delete',
+  meta: {
+    docs: {
+      description: 'Disallow delete operator for arrays',
+      recommended: 'error',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      arrayDeleteViolation:
+        'Using delete operator for arrays are forbidden. Use array.splice instead.',
+    },
+    schema: [],
+    type: 'problem',
+    fixable: 'code',
+  },
+  defaultOptions: [],
+  create(context) {
+    return {
+      "UnaryExpression[operator='delete']"(
+        node: TSESTree.UnaryExpression,
+      ): void {
```

Comment:
We can get even more precise with the query here:

```suggestion
      "UnaryExpression[operator='delete'] MemberExpression[computed]"(
        node: TSESTree.MemberExpressionComputedName & { parent: TSESTree.UnaryExpression },
      ): void {
```

Two main benefits:
* It gives more precision to the type system on what the node shapes are
* It succinctly filters away undesirable node shapes that this rule wouldn't want to check for, like a delete on a non-computed member expression (`delete a.b`)

That first benefit is quite nice because it'll save the rule from having to call `getChildAt`. We generally prefer not to use that API because it's imprecise and doesn't give much information to the type system on the types of the children. Instead of `expression.getChildAt(0)`, with this suggested change, you can ask for `node.object` (roughly).

Overall: instead of looking at `originalNode` and accessing children from it, try to stay within the ESTree AST -- only dipping into TypeScript-land when you're sure you'll need type information.